File: katas/BowlingGame/js/src/Game.ts
export class Game {
  private rolls: number[] = [];
  private currentRoll: number = 0;

  public roll(numPinsHit: number) {
    this.rolls[this.currentRoll] = numPinsHit;
    this.currentRoll++;
  }

  public calculateScore() {
    let calculatedScore: number = 0;
    let frameIndex = 0;
    for (let frame = 0; frame < 10; frame++) {
      if (this.isStrike(frameIndex)) {
        calculatedScore += 10 + this.calculateStrikeBonus(frameIndex);
        frameIndex += 1;
      } else if (this.isSpare(frameIndex)) {
        calculatedScore += 10 + this.calculateSpareBonus(frameIndex);
        frameIndex += 2;
      } else {
        calculatedScore += this.sumRollsInFrame(frameIndex);
        frameIndex += 2;
      }
    }
    
    return calculatedScore;
  }

  isSpare(frameIndex: number) {
    return (this.rolls[frameIndex] + this.rolls[frameIndex + 1]) === 10;
  }

  isStrike(frameIndex: number) {
    return this.rolls[frameIndex] === 10;
  }

  sumRollsInFrame(frameIndex: number) {
    return this.rolls[frameIndex] + this.rolls[frameIndex + 1];
  }

  calculateSpareBonus(frameIndex: number) {
    const firstRollInNextFrame = this.rolls[frameIndex + 2];
    return firstRollInNextFrame;
  }

  calculateStrikeBonus(frameIndex: number) {
    const firstRollAfterStrike = this.rolls[frameIndex + 1];
    const secondRollAfterStrike = this.rolls[frameIndex + 2];

    return firstRollAfterStrike + secondRollAfterStrike;
  }
}

File: katas/BowlingGame/ruby/src/Game.rb
class Game

  def initialize
    @current_roll = 0
    @rolls = []
  end

  def roll(pins)
    @rolls[@current_roll] = pins
    @current_roll += 1
  end

  def score
    result = 0
    frame_index = 0

    for frame in 0 ... 10
      if is_strike?(frame_index)
        result += 10 + calculate_strike_bonus(frame_index)
        frame_index += 1
      elsif is_spare?(frame_index)
        result += 10 + calculate_spare_bonus(frame_index)
        frame_index += 2
      else
        result += sum_rolls_in_frame(frame_index)
        frame_index += 2
      end
    end

    result
  end

  def calculate_strike_bonus(frame_index)
    @rolls[frame_index + 1] + @rolls[frame_index + 2]
  end

  def calculate_spare_bonus(frame_index)
    @rolls[frame_index + 2]
  end

  def sum_rolls_in_frame(frame_index)
    @rolls[frame_index] + @rolls[frame_index + 1]
  end

  def is_strike?(frame_index)
    @rolls[frame_index] == 10
  end

  def is_spare?(frame_index)
    @rolls[frame_index] + @rolls[frame_index + 1] == 10
  end
end
